pkg/cni/plugin.go

```diff
@@ -0,0 +1,259 @@
+				return fmt.Errorf("failed to add hardware addr to %q: %v", ifName, err)
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	macvtap := &current.Interface{Name: ifName}
+	err = configureArp(macvtapIface, netns)
+	if err != nil {
+		return nil, err
+	}
+
+	err = renameMacvtapIface(macvtapIface, macvtap, ifName, netns)
+	if err != nil {
+		return nil, err
+	}
+
+	return macvtap, err
+}
+
+func renameMacvtapIface(macvtapLink netlink.Link, macvtapIface *current.Interface, ifaceName string, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		err := ip.RenameLink(macvtapLink.Attrs().Name, ifaceName)
+		if err != nil {
+			_ = netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to rename macvlan to %q: %v", ifaceName, err)
+		}
+
+		updatedLink := macvtapLink
+		updatedLink.Attrs().Name = ifaceName
+
+		// Re-fetch macvlan to get all properties/attributes
+		contMacvlan, err := netlink.LinkByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("failed to refetch macvlan %q: %v", ifaceName, err)
+		}
+		macvtapIface.Mac = contMacvlan.Attrs().HardwareAddr.String()
+		macvtapIface.Sandbox = netns.Path()
+
+		return nil
+	})
+	return err
+}
+
+// CmdAdd - CNI interface
+func CmdAdd(args *skel.CmdArgs) error {
+	var err error
+	netConf, cniVersion, err := loadConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	envArgs, err := getEnvArgs(args.Args)
+	if err != nil {
+		return err
+	}
+
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		return fmt.Errorf("failed to open netns %q: %v", netns, err)
+	}
+	defer netns.Close()
+
+	var mac net.HardwareAddr
+	if envArgs.MAC != "" {
+		mac, err = net.ParseMAC(string(envArgs.MAC))
+		if err != nil {
+			return err
+		}
+	}
+
+	macvtapInterface, err := configureMacvtap(netConf, mac, args.IfName, netns)
+	if err != nil {
+		return err
+	}
+
+	// Delete link if err to avoid link leak in this ns
+	defer func() {
+		if err != nil {
+			netns.Do(func(_ ns.NetNS) error {
+				return ip.DelLinkByName(args.IfName)
+			})
+		}
+	}()
+
+	err = netns.Do(func(_ ns.NetNS) error {
+		macvtapIface, err := netlink.LinkByName(args.IfName)
+		if err != nil {
+			return fmt.Errorf("failed to lookup new macvtapdevice %q: %v", args.IfName, err)
+		}
+
+		if err := netlink.LinkSetUp(macvtapIface); err != nil {
+			return fmt.Errorf("failed to set macvtap iface up: %v", err)
+		}
+		return nil
+	})
+
+	result := &current.Result{
+		CNIVersion: cniVersion,
+		Interfaces: []*current.Interface{macvtapInterface},
+	}
+
+	return types.PrintResult(result, cniVersion)
+}
+
+// CmdDel - CNI plugin Interface
+func CmdDel(args *skel.CmdArgs) error {
+	if args.Netns == "" {
+		return nil
+	}
+
+	// There is a netns so try to clean up. Delete can be called multiple times
+	// so don't return an error if the device is already removed.
+	err := ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+
+		if err := ip.DelLinkByName(args.IfName); err != nil {
```

Comment:
Use `LinkDelete` @ `pkg/util/netlink.go`

Reply:
Cleanup is deferred until the end of CmdAdd.